### PR TITLE
Send API

### PR DESCRIPTION
**Pull request by somdoron**
New methods documentation is still missing. I will send another PR soon with the documentation

### CODE REVIEW

File: src/NetMQ.Tests/SocketTests.cs [hunk 1]
```diff
@@ -552,9 +568,7 @@ public void ASubscriberSocketThatGetDisconnectedBlocksItsContextFromBeingDispose
             using (var pub = pubContext.CreatePublisherSocket())
             using (var sub = subContext.CreateSubscriberSocket())
             {
-                pub.Options.Linger = TimeSpan.FromSeconds(0);
-                pub.Options.SendTimeout = TimeSpan.FromSeconds(2);
-
+                pub.Options.Linger = TimeSpan.FromSeconds(0);                
                 sub.Options.Linger = TimeSpan.FromSeconds(0);
```
Comment (drewnoakes) [inline src/NetMQ.Tests/SocketTests.cs]: Is this supposed to be the same line twice?

File: src/NetMQ/SendReceiveConstants.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Text;
+
+namespace NetMQ
+{
+    public static class SendReceiveConstans
```
Comment (drewnoakes) [inline src/NetMQ/SendReceiveConstants.cs]: Typo here: `SendReceiveConstants`

File: src/NetMQ/OutgoingSocketExtensions.cs [hunk 3]
```diff
@@ -98,8 +200,182 @@ public static IOutgoingSocket SendMore([NotNull] this IOutgoingSocket socket, [N
...
+        #region Blocking
+
+        public static void SendMultipartBytes([NotNull] this IOutgoingSocket socket, params byte[][] frames)
+        {
+            SendMultipartBytes(socket, (IEnumerable<byte[]>)frames);
+        }
+
+        public static void SendMultipartBytes([NotNull] this IOutgoingSocket socket, IEnumerable<byte[]> frames)
+        {
+            var enumerator = frames.GetEnumerator();
```
Comment (drewnoakes) [inline src/NetMQ/OutgoingSocketExtensions.cs]: Note sure it matters much, but I think .NET calls `Dispose` on any `IEnumerator<T>` which is `IDisposable`. Info on [MSDN](https://msdn.microsoft.com/en-us/library/aa664754(v=vs.71).aspx).

File: src/NetMQ/OutgoingSocketExtensions.cs [hunk 4]
```diff
@@ -98,8 +200,182 @@ public static IOutgoingSocket SendMore([NotNull] this IOutgoingSocket socket, [N
...
+
+        public static bool TrySendMultipartBytes([NotNull] this IOutgoingSocket socket, TimeSpan timeout, params byte[][] frames)
+        {
+            return TrySendMultipartBytes(socket, timeout, (IEnumerable<byte[]>)frames);
+        }
+
+        public static bool TrySendMultipartBytes([NotNull] this IOutgoingSocket socket, TimeSpan timeout,
+            IEnumerable<byte[]> frames)
+        {
+            var enumerator = frames.GetEnumerator();
```
Comment (drewnoakes) [inline src/NetMQ/OutgoingSocketExtensions.cs]: Note sure it matters much, but I think .NET calls `Dispose` on any `IEnumerator<T>` which is `IDisposable`. Info on [MSDN](https://msdn.microsoft.com/en-us/library/aa664754(v=vs.71).aspx).

**Comment by drewnoakes** [comment]
Looks great. I added a few minor comments inline.

Also, did `ConsoleApplication1` and `ConsoleApplication2` projects get added by accident?

File: src/NetMQ/NetMQActor.cs [hunk 5]
```diff
@@ -229,17 +229,11 @@ private void RunShim()
...
-
-            // Do not block, if the other end of the pipe is already deleted
-            m_shim.Options.SendTimeout = TimeSpan.Zero;
-
-            try
             {
-                m_shim.SignalOK();
             }
-            catch (AgainException)
-            {}
```
Comment (drewnoakes) [inline src/NetMQ/NetMQActor.cs]: Should `AgainException` be marked obsolete? Client code may be updated to use the new API, yet still have `catch (AgainException)` blocks.

File: src/NetMQ/OutgoingSocketExtensions.cs (hunk 3 above)
Comment (somdoron) [inline src/NetMQ/OutgoingSocketExtensions.cs]: But is is an IEnumerable, not IEnumerator

File: src/NetMQ/NetMQActor.cs (hunk 5 above)
Comment (somdoron) [inline src/NetMQ/NetMQActor.cs]: yes, good idea

**Comment by somdoron** [comment]
I will fix your comments and add the docs soon

File: src/NetMQ/OutgoingSocketExtensions.cs (hunk 3 above)
Comment (drewnoakes) [inline src/NetMQ/OutgoingSocketExtensions.cs]: C# `foreach` disposes the IEnumerator if it implements IDisposable. I'm not sure it's important, but as this API is public it's theoretically possible client code could depend upon this behaviour.

```csharp
E enumerator = (collection).GetEnumerator();
try {
   while (enumerator.MoveNext()) {
      ElementType element = (ElementType)enumerator.Current;
      statement;
   }
}
finally {
   IDisposable disposable = enumerator as System.IDisposable;
   if (disposable != null) disposable.Dispose();
}
```

**Comment by somdoron** [comment]
I rebased to fix the type and remove the console application.
Also:
* Complete xml comments
* Obsolete again exception
* Fix enumerator

**Pull request merged**